src/utils.cpp

```diff
@@ -764,5 +770,216 @@ bool executePreAction(const nlohmann::json& i_parsedConfigJson,
 
     return true;
 }
+
+std::string getTruncatedInvPath(const std::string& i_invPath)
+{
+    // Truncate /xyz/openbmc_project/inventory if present
+    std::string l_truncatedInvPath = i_invPath;
+    std::string::size_type l_pos = i_invPath.find(constants::pimPath);
+
+    if (l_pos != std::string::npos)
+    {
+        l_truncatedInvPath = i_invPath.substr(l_pos + 30);
+    }
+    return l_truncatedInvPath;
+}
+
+bool isInventoryPathFromJSON(const types::Path i_path,
+                             const nlohmann::json i_jsonObj)
+{
+    if (!i_jsonObj.contains("frus"))
+    {
+        return false;
+    }
+
+    // Revisit: If JSON has complete path with prefix
+    // /xyz/openbmc_project/inventory
+    auto l_truncatedPath = utils::getTruncatedInvPath(i_path);
+
+    const nlohmann::json& l_listOfFrus =
+        i_jsonObj["frus"].get_ref<const nlohmann::json::object_t&>();
+
+    for (const auto& l_itemFRUS : l_listOfFrus.items())
+    {
+        for (const auto& l_fru : l_itemFRUS.value())
+        {
+            if (l_fru.find("inventoryPath") != l_fru.end() &&
+                l_truncatedPath == l_fru.value("inventoryPath", ""))
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool isDbusInvPath(const std::string& i_path)
+{
+    if (i_path.find(constants::pimPath) != std::string::npos)
+    {
+        return true;
+    }
+    return false;
+}
+
+types::DbusVariantType
+    readKeywordFromHardware(const types::Path i_path,
+                            const types::ReadVpdParams i_data,
+                            const nlohmann::json i_jsonObj)
+{
+    auto l_vpdStartOffset =
+        utils::getVPDOffset(i_jsonObj, getPrimaryPath(i_jsonObj, i_path));
```

Comment:
> getPrimaryPath


Are we calling this API for all the paths blindly? If yes then why so? If the the path is already a primary path why do we need to go through the loop unnecessarily? Is there a way to optimize that? Please look into that.

Reply:
Check for a primary path can be done without looping around the JSON. The implementation here can be optimized.